recipes-app/iot2050-firmware-update/files/iot2050-firmware-update

```diff
@@ -312,15 +340,66 @@ class ManagedFirmwareUpdate(FirmwareUpdate):
             while choice > len(firmware_names) or choice < 1:
                 print("Out of range, please reinput your choice:")
                 choice = int(input("-> "))
-            
+
             firmware_name = firmware_names[choice - 1]
 
         print("Will write {} to the SPI flash".format(firmware_name))
 
         self.firmware.seek(0)
-        self.firmware = tarfile.open(fileobj=self.firmware).extractfile(firmware_name)
-        super().update_firmware()
+        with tarfile.open(fileobj=self.firmware) as f:
+            self.firmware = f.extractfile(firmware_name)
+            f.extract(member=self.uboot_default_env_file, path='.')
+            super().update_firmware()
+
+    def update_uboot_env(self, env_list):
+        '''
+        restore uboot env
+        retrive the uboot default env file belong to the upgrade firmware.
+        makeup the restore env binary and flash to the env mtd partition
+        '''
+        env_default_binary = 'uEnv_default.bin'
+        # /dev/mtd3     /dev/mtd4
+        env_partition_list = [3, 4]
+        # append value to the end
+        with open(self.uboot_default_env_file, encoding="utf-8", mode="a") as file:
+            for value in env_list:
+                file.write(value)
+                file.write("\n")
+
+        for mtd_num in env_partition_list:
+            mtd_dev_path, mtd_size, mtd_erasesize, mtd_name = self.get_mtd_info(
+                mtd_num)
+            subprocess.call('mkenvimage -s {} -r -o {} {}'.format(mtd_size,
+                            env_default_binary, self.uboot_default_env_file), shell=True)
```

Comment:
Use `subprocess.run()` instead, and for the command, do not use the `-o {}` but direct to the `stdout` and capture it from the python script. Also check the `stderr` for robust (the `mkenvimage` command is not guaranteed to always work)

Reply:
Just as what you've done below with `subprocess.check_output`.

However, I am not sure if this works with the binary stream. You can have a try.

Reply:
use `subprocess.run()` API to handle the command and add exception mechanism